File: Problem Patterns/Math/valid-perfect-square.py
# Problem Link: https://leetcode.com/problems/valid-perfect-square

class Solution:
    def isPerfectSquare(self, num: int) -> bool:

        if num == 1:
            return True
        
        lo, hi = 0, num

        while lo <= hi:

            mid = (hi + lo) // 2
            if mid == hi or mid == lo:
                break

            # print(lo, hi, mid, mid * mid,  num)/

            if mid * mid == num:
                return True
            elif mid * mid > num:
                hi = mid
            else:
                lo = mid

        return False

File: Problem Patterns/Math/valid-perfect-square.java
# Problem Link: https://leetcode.com/problems/valid-perfect-square/

class Solution {
        public boolean isPerfectSquare(int num) {
        if (num == 1) {
            return true;
        }

        long lo = 0, hi = num;

        while (lo <= hi) {
            long mid = (hi + lo) / 2;
            if (mid == hi || mid == lo) {
                break;
            }

            if (mid * mid == num) {
                return true;
            } else if (mid * mid > num) {
                hi = mid;
            } else {
                lo = mid;
            }
        }

        return false;
    } 
}
